### PR TITLE
CEDS-1478 Add Pagination & Search by Status to Find All Decs by EORI

### PR DESCRIPTION
**Pull request by ed0906**

### CODE REVIEW

File: app/uk/gov/hmrc/exports/models/DeclarationSearch.scala [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.exports.models
+
+import play.api.libs.json.{Json, OFormat}
+import uk.gov.hmrc.exports.models.declaration.DeclarationStatus.DeclarationStatus
+
+case class DeclarationSearch(eori: String, status: Option[DeclarationStatus] = None)
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/models/DeclarationSearch.scala]: From experience we usually end up adding more and more search filters over time so I've stuck it in a Search class for now

File: app/uk/gov/hmrc/exports/models/Page.scala [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+package uk.gov.hmrc.exports.models
+
+import play.api.libs.json.{Json, OFormat}
+import play.api.mvc.QueryStringBindable
+
+case class Page(index: Int = 1, size: Int = 50)
+
+object Page {
+  implicit val format: OFormat[Page] = Json.format[Page]
+  implicit def bindable(implicit intBinder: QueryStringBindable[Int]): QueryStringBindable[Page] =
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/models/Page.scala]: If you haven't seen these before it means we can use complex types in `app.routes` and  Controller params and it "just works"

I expect Notifications will want this eventually too

File: app/uk/gov/hmrc/exports/models/Paginated.scala [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+package uk.gov.hmrc.exports.models
+
+import play.api.libs.json._
+
+case class Paginated[T](results: Seq[T], page: Page, total: Long)
+object Paginated {
+  def apply[T](results: T*): Paginated[T] = Paginated[T](results, Page(), results.size)
+  def empty[T](page: Page) = Paginated(Seq.empty[T], page, 0)
+
+  implicit def writes[T](implicit fmt: Writes[T]): Writes[Paginated[T]] = new Writes[Paginated[T]] {
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/models/Paginated.scala]: Generics means doing custom `Writes`....

File: app/uk/gov/hmrc/exports/models/Page.scala (hunk 2 above)
Comment (bambuchaAdm) [reply app/uk/gov/hmrc/exports/models/Page.scala]: I don't think we have to take implicit argument here - this `QueryStringBindable[Int]` could be field in `QueryStringBindable[Page]` with usage of `implicitly` language function

File: app/uk/gov/hmrc/exports/models/Page.scala [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    new QueryStringBindable[Page] {
+      private def queryParamIndex(key: String) = key + "-index"
+      private def queryParamSize(key: String) = key + "-size"
+
+      override def bind(key: String, params: Map[String, Seq[String]]): Option[Either[String, Page]] = {
+        val index: Option[Int] = params.get(queryParamIndex(key)).flatMap(_.headOption).map(_.toInt)
+        val size: Option[Int] = params.get(queryParamSize(key)).flatMap(_.headOption).map(_.toInt)
+        (index, size) match {
```
Comment (bambuchaAdm) [inline app/uk/gov/hmrc/exports/models/Page.scala]: maybe just getOrElse on options instead of matching ?

Comment (ed0906) [reply app/uk/gov/hmrc/exports/models/Page.scala]: I kind of wanted the default values to live inside `Page` so we only have them in one place

Comment (ed0906) [reply app/uk/gov/hmrc/exports/models/Page.scala]: I guess I could make them fields that I can access from here

Comment (ed0906) [reply app/uk/gov/hmrc/exports/models/Page.scala]: Fixed

File: app/uk/gov/hmrc/exports/models/Page.scala (hunk 2 above)
Comment (ed0906) [reply app/uk/gov/hmrc/exports/models/Page.scala]: Nice

Comment (ed0906) [reply app/uk/gov/hmrc/exports/models/Page.scala]: Fixed

**Pull request merged**